File: apps/api/app/schemas/goals.py
"""Schemas for the goals endpoints."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator

GoalType = Literal[
    "weight_loss",
    "weight_gain",
    "body_fat",
    "strength",
    "endurance",
    "habit",
    "custom",
]

GoalStatus = Literal["active", "completed", "paused", "cancelled"]


# ── Requests ──────────────────────────────────────────────────────────────────

class CreateGoalRequest(BaseModel):
    goal_type: GoalType
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=1000)

    # Numeric tracking — all optional.  If any numeric field is provided,
    # target_unit should also be provided (validated below).
    starting_value: float | None = Field(default=None, ge=0)
    target_value: float | None = Field(default=None, ge=0)
    current_value: float | None = Field(default=None, ge=0)
    target_unit: str | None = Field(default=None, max_length=32)

    deadline: date | None = None

    @model_validator(mode="after")
    def numeric_fields_consistent(self) -> CreateGoalRequest:
        has_value = any(
            v is not None
            for v in (self.starting_value, self.target_value, self.current_value)
        )
        if has_value and not self.target_unit:
            raise ValueError(
                "target_unit is required when any numeric value is provided"
            )
        return self


class UpdateGoalRequest(BaseModel):
    """All fields optional — send only what changed."""

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    goal_type: GoalType | None = None
    starting_value: float | None = Field(default=None, ge=0)
    target_value: float | None = Field(default=None, ge=0)
    current_value: float | None = Field(default=None, ge=0)
    target_unit: str | None = Field(default=None, max_length=32)
    deadline: date | None = None
    status: GoalStatus | None = None


# ── Response ──────────────────────────────────────────────────────────────────

class GoalResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    user_id: uuid.UUID
    goal_type: str
    title: str
    description: str | None
    starting_value: float | None
    target_value: float | None
    current_value: float | None
    target_unit: str | None
    deadline: date | None
    status: str
    completed_at: datetime | None
    is_public: bool
    created_at: datetime
    updated_at: datetime

    # Computed — not stored in DB.
    progress_pct: float | None = None


class GoalListResponse(BaseModel):
    goals: list[GoalResponse]
    total: int
    page: int
    page_size: int
    has_next: bool

File: apps/web/src/types/goals.ts
/**
 * TypeScript types for the goals domain.
 * Mirror the GoalResponse and list schema from the FastAPI backend.
 */

export const GOAL_TYPES = [
  "weight_loss",
  "weight_gain",
  "body_fat",
  "strength",
  "endurance",
  "habit",
  "custom",
] as const;

export type GoalType = (typeof GOAL_TYPES)[number];

export const GOAL_TYPE_LABELS: Record<GoalType, string> = {
  weight_loss: "Weight Loss",
  weight_gain: "Weight Gain",
  body_fat: "Body Fat",
  strength: "Strength",
  endurance: "Endurance",
  habit: "Habit",
  custom: "Custom",
};

export const GOAL_STATUSES = [
  "active",
  "paused",
  "completed",
  "cancelled",
] as const;

export type GoalStatus = (typeof GOAL_STATUSES)[number];

/** Allowed transitions from a given status. */
export const ALLOWED_STATUS_TRANSITIONS: Record<GoalStatus, GoalStatus[]> = {
  active: ["paused", "completed", "cancelled"],
  paused: ["active", "cancelled"],
  completed: [],
  cancelled: [],
};

export interface Goal {
  id: string;
  user_id: string;
  goal_type: GoalType;
  title: string;
  description: string | null;
  starting_value: number | null;
  target_value: number | null;
  current_value: number | null;
  target_unit: string | null;
  deadline: string | null; // ISO date string, e.g. "2025-12-31"
  status: GoalStatus;
  completed_at: string | null;
  is_public: boolean;
  progress_pct: number | null; // computed by the backend, 0-100 or null
  created_at: string;
  updated_at: string;
}

export interface GoalListResponse {
  goals: Goal[];
  total: number;
  page: number;
  page_size: number;
  has_next: boolean;
}

export interface CreateGoalPayload {
  goal_type: GoalType;
  title: string;
  description?: string;
  starting_value?: number;
  target_value?: number;
  current_value?: number;
  target_unit?: string;
  deadline?: string;
}

export interface UpdateGoalPayload {
  title?: string;
  description?: string | null;
  goal_type?: GoalType;
  starting_value?: number | null;
  target_value?: number | null;
  current_value?: number | null;
  target_unit?: string | null;
  deadline?: string | null;
  status?: GoalStatus;
}
